src/utils.py

```diff
@@ -232,92 +238,98 @@ def wrap_atom_predicates(atoms: Collection[LiftedOrGroundAtom],
             actions.append(act)
             states.append(state)
-            if task.goal.issubset(atoms):
-                goal_reached = True
+            if termination_function(state):
+                term_fn_hit = True
                 break
-    if make_video:
-        assert render is not None
-        # Explanation of type ignore: mypy currently does not
-        # support Callables with optional arguments. mypy
-        # extensions does, but for the sake of avoiding an
-        # additional dependency, we'll just ignore this here.
-        video.extend(render(state, task))  # type: ignore
-    if annotate_traj_with_goal:
-        traj = LowLevelTrajectory(states, actions, task.goal)
-    else:
-        traj = LowLevelTrajectory(states, actions)
+            # Detect if stuck; skip potentially expensive simulation.
+            if state.allclose(last_state):
+                break
+    traj = LowLevelTrajectory(states, actions)
+    return traj, term_fn_hit
+
+
+def run_policy_on_task(
+    policy: Callable[[State], Action],
+    task: Task,
+    simulator: Callable[[State, Action], State],
+    predicates: Collection[Predicate],
+    max_num_steps: int,
+    render: Optional[Callable[[State, Task, Optional[Action]],
+                              List[Image]]] = None,
+) -> Tuple[LowLevelTrajectory, Video, bool]:
+    """A light wrapper around run_policy_until that takes in a task and uses
+    achieving the task's goal as the termination_function.
+
+    Returns the trajectory and whether it achieves the task goal. Also
+    optionally returns a video, if a render function is provided.
+    """
+
+    def _terminal(state: State) -> bool:
+        atoms = abstract(state, predicates)
+        return task.goal.issubset(atoms)
+
+    traj, goal_reached = run_policy_until(policy, simulator, task.init,
+                                          _terminal, max_num_steps)
+    video: Video = []
+    if render is not None:  # step through the traj again, making the video
+        for i, state in enumerate(traj.states):
+            act = traj.actions[i] if i < len(traj.states) - 1 else None
+            video.extend(render(state, task, act))
     return traj, video, goal_reached
 
 
 def policy_solves_task(policy: Callable[[State], Action], task: Task,
                        simulator: Callable[[State, Action], State],
                        predicates: Collection[Predicate]) -> bool:
-    """Return whether the given policy solves the given task."""
+    """A light wrapper around run_policy_on_task that returns whether the given
+    policy solves the given task."""
     _, _, solved = run_policy_on_task(policy, task, simulator, predicates,
                                       CFG.max_num_steps_check_policy)
     return solved
 
 
-def option_to_trajectory(init: State, simulator: Callable[[State, Action],
-                                                          State],
+def option_to_trajectory(init_state: State,
```

Comment:
maybe we can actually get rid of this function altogether after you change the thing i commented above? users would need to make sure that the option is initiable, but i'm guessing that's already happening...